Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm

```diff
@@ -121,19 +182,40 @@
         parsedDetails.type = parsedType;
     }
 
-    MockBookmarkNode newNode;
-    newNode.id = bookmark[idKey];
-    newNode.url = bookmark[urlKey];
-    newNode.title = bookmark[titleKey];
-    newNode.parentId = bookmark[parentIdKey];
-    newNode.index = parsedDetails.index.value();
-    newNode.type = parsedDetails.type.value_or(WebExtensionAPIBookmarks::BookmarkTreeNodeType::Folder);
+    Ref<MockBookmarkNode> newNode = MockBookmarkNode::create();
+    newNode->id = bookmark[idKey];
+    newNode->url = bookmark[urlKey];
+    newNode->title = bookmark[titleKey];
+    newNode->parentId = bookmark[parentIdKey] ? bookmark[parentIdKey] : bookmarksRootId;
+    newNode->index = parsedDetails.index.value();
+    newNode->type = bookmark[urlKey] ? WebExtensionAPIBookmarks::BookmarkTreeNodeType::Bookmark : WebExtensionAPIBookmarks::BookmarkTreeNodeType::Folder;
 
     double msSinceEpoch = dynamic_objc_cast<NSNumber>(bookmark[dateAddedKey]).doubleValue;
-    newNode.dateAdded = WTF::WallTime::fromRawSeconds(msSinceEpoch / 1000.0);
+    newNode->dateAdded = WTF::WallTime::fromRawSeconds(msSinceEpoch / 1000.0);
+
+    String parentId = newNode->parentId;
+    String newNodeId = newNode->id;
+
+    auto parentOptional = m_mockBookmarks.getOptional(parentId);
+    if (!parentOptional) {
+        *outExceptionString = toErrorString(nullString(), parentIdKey, @"parentId could not be mapped to a node").createNSString().autorelease();
```

Comment:
```suggestion
        *outExceptionString = toErrorString(nullString(), parentIdKey, @"it could not be mapped to a node").createNSString().autorelease();
```